thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js

```diff
@@ -0,0 +1,303 @@
+import Ember from 'ember';
+import _ from 'lodash';
+
+//
+// Config
+//
+
+const EVENT_TABLE_COLUMNS = [
+  {
+    template: 'custom/checkbox',
+    useFilter: false,
+    mayBeHidden: false,
+    className: 'events-table__column--checkbox'
+  },
+  {
+    propertyName: 'label',
+    title: 'Event Name',
+    className: 'events-table__column'
+  },
+  {
+    propertyName: 'eventType',
+    title: 'Type',
+    filterWithSelect: true,
+    sortFilterOptions: true,
+    className: 'events-table__column--compact'
+  },
+  {
+    propertyName: 'start',
+    title: 'Start',
+    className: 'events-table__column--compact',
+    disableFiltering: true
+  },
+  {
+    propertyName: 'end',
+    title: 'End',
+    className: 'events-table__column--compact',
+    disableFiltering: true
+  }
+];
+
+//
+// Controller
+//
+
+export default Ember.Controller.extend({
+  selectedUrns: null, // Set
+
+  filteredUrns: null, // Set
+
+  invisibleUrns: null, // Set
+
+  _timeseriesCache: null, // {}
+
+  _entitiesCache: null, // {}
+
+  _pendingRequests: null, // {}
+
+  entities: Ember.computed(
+    '_entitiesLoader',
+    '_entitiesCache',
+    function () {
+      console.log('entities()');
+      this.get('_entitiesLoader'); // trigger loader. hacky
+      return this.get('_entitiesCache') || {};
+    }
+  ),
+
+  timeseries: Ember.computed(
+    '_timeseriesLoader',
+    '_timeseriesCache',
+    function () {
+      console.log('timeseries()');
+      this.get('_timeseriesLoader'); // trigger loader. hacky
+      return this.get('_timeseriesCache') || {};
+    }
+  ),
+
+  chartSelectedUrns: Ember.computed(
+    'model.primaryMetricUrn',
+    'selectedUrns',
+    'invisibleUrns',
+    function () {
+      console.log('chartSelectedUrns()');
+      const selectedUrns = this.get('selectedUrns');
+      const invisibleUrns = this.get('invisibleUrns');
+      const primaryMetricUrn = this.get('model.primaryMetricUrn');
+
+      // console.log('chartSelectedUrns: selectedUrns', selectedUrns);
+      // console.log('chartSelectedUrns: invisibleUrns', invisibleUrns);
+      // console.log('chartSelectedUrns: primaryMetricUrn', primaryMetricUrn);
+      //
+      const output = new Set(selectedUrns);
+      output.add(primaryMetricUrn);
+      invisibleUrns.forEach(urn => output.delete(urn));
+
+      return output;
+    }
+  ),
+
+  eventTableEntities: Ember.computed(
+    'entities',
```

Comment:
Do we care about entities as a whole (reference to the array), or do we also want to recompute when properties changes on entities?  Maybe https://guides.emberjs.com/v2.16.0/object-model/computed-properties-and-aggregate-data/#toc_code-each-code might be helpful

Reply:
entities are immutable for the frontend